File: code/stack/cpp/646.c
#include "Stack.h"

Stack::Stack()
{
  top=-1;  
  size=5; 
  p = new int[size]; 
              
}

Stack::Stack(int s)
{
  size=s; 
  top=-1;
  p=new int[s]; 
                 
}

bool Stack::isEmpty()
{
  if(top==-1)
  {
   return true;
   }
   else 
   return false;   
              
}

void Stack::push(int value) 
{  
  if(top<size)
  {
  top++;
  p[top]=value;
  }
  else
  {
  int* temp = new int[2*size];
  for(int i=0;i<size;i++)
    {
    temp[i]=p[i];
    }
   delete[] p;
   p=temp;
   }
}


int Stack::pop()
{
 return p[top--];           
}


                 

File: code/stack/java/646.java
package com.gems.algorithms;

public class Stack<T> {

	private T[] stack; 
	private int N;
	
	public Stack(int size)
	{
		stack = (T[]) new Object[size];
		N = 0;
	}
	
	public boolean isEmpty()
	{
		return 0 == N;
	}
	
	public int size()
	{
		return stack.length;
	}
	
	private void resize(int size)
	{
		T[] array = (T[]) new Object[size];
		for (int i=0; i<stack.length; i++) 
			array[i] = stack[i];
		stack = array;
	}
	
	public void push(T item)
	{
		if (N == stack.length) resize(2*stack.length);
		stack[N++] = item;
	}
	
	public T pop() 
	{
		T item = stack[--N];
		stack[N] = null;
		if (N > 100 && N == stack.length / 4) resize(stack.length / 2);
		return item;
	}
}
